ser/train.py

```diff
@@ -0,0 +1,64 @@
+import torch
+import torch.nn.functional as F
+
+def train_model(validation_dataloader, training_dataloader, model, optimizer, epochs, device, name):
+    # train
+    with open('./experiments/model_'+name+'.txt', 'w') as f:
+                    f.write(f'Experiment file for model: {name} \n Model architecture: {model}')
+    best_acc = 0
+    best_acc_epoch = 0
+    best_model = None
+    best_batch = 0
+    for epoch in range(epochs):
+        for i, (images, labels) in enumerate(training_dataloader):
+            images, labels = images.to(device), labels.to(device)
+            model.train()
+            optimizer.zero_grad()
+            output = model(images)
+            loss = F.nll_loss(output, labels)
+            loss.backward()
+            optimizer.step()
+
+            update_message = (
+                f"\nTrain Epoch: {epoch} | Batch: {i}/{len(training_dataloader)}"
+                f"| Loss: {loss.item():.4f}"
+            )
+
+            print(update_message)
+            with open('./experiments/model_'+name+'.txt', 'a') as f:
+                    f.write(update_message)
+        
+
+            # validate
+            val_loss = 0
+            correct = 0
+            with torch.no_grad():
```

Comment:
So the performance issues I mentioned near teh start is still here :). The validation algorithm should be run once per epoch instead of once per batch...

Part of the reason this is hard to spot is that the training/validation algorithms are all inline in one pretty function definition, so it's quite hard to see that it's indented wrong.

If you instead factor out training and validation into two separate functions, the bug will just go as way, as well as making the code clearer. I.e. this would be something like:

```for epoch in epochs:
    train()
    validate()
```